### ISSUE TITLE
Audience screen invisible in KDE Plasma 5

### ISSUE THREAD

**Issue by oskar456** [pdfpc/pdfpc#346]
This is a followup to issue #221. Any pdfpc version > 4.0.5 does not show the audience screen on the secondary monitor when using KDE Plasma 5 desktop environment. The only workaround I've figured out is to downgrade to version 4.0.5. Any help would be much appreciated.

**Comment by mmbcorley** [comment]
Seconding this.  I simply can't make pdfpc work in KDE Plasma 5 -- interested in any workarounds people may have found.

**Comment by 1lm** [comment]
Encouraged by older Gtk/GDI interface versions, pdfpc equates a "monitor" with a "screen". But with plasma, there may be only one screen (the workspace) which may extend over two or more monitors. Hence pdfpc will not work correctly, in turn because "screen_number" will be silently disregarded (without triggering an error message). Thus the presentation window will still be created, but will be hidden behind the full screen presenter window. 

With Plasma, the problem could be fixed by using an appropriate call to window.move(), but even this is not a viable solution when using a wayland compositor. There, window.move() will be disregarded since the wayland design philosophy dictates that the placement of windows must be in the sole responsibility of the window manager.

There is a workaround: use "alt-tab" or whatever key combination to reach the window-manager's context menu of the presenter and the presentation window in the taskbar (or, I believe, "ctrl-shift-F3" with wayland to reach this menu for any particular window directly, even in full screen mode), cancel the full screen mode and use the mouse to move the window, or send the presentation window directly to the other monitor by means of the appropriate context menu entry.

The longer I think about it the more I believe that the most flexible solution of this problem would consist in starting both windows with a moderate size in a decorated form (with title bar and window manager's context menus), so everybody can simply use the mouse to place the windows on the right monitor and maximize them there. This would work with any number of monitors and could be also automated by applying special rules to pdfpc's windows using the window rules tab in Plasma's system settings dialog.

**Comment by mmbcorley** [comment]
@1lm thanks for the workaround!  Tested, and works (after assigning a shortcut to "move to next screen" in plasma).

**Comment by 1lm** [comment]
Just found that alt-F3 is the shortcut to reach the window manager's context menu within Plasma. So use alt-tab to bring the presentation window to the foreground and then press alt-F3 together to display the context menu.

**Comment by fnevgeny** [comment]
Please try the "multimonitor" branch.

**Comment by oskar456** [comment]
It works for me in KDE Plasma 5.13.5. I had to compile it with vala 0.36.13, as with vala-0.32, compilation fails on

    The name `get_monitor' does not exist in the context of `Gdk.Display?'

With vala 0.36, I only get just this deprecation warning:

    pdfpc/src/classes/window/fullscreen.vala:108.21-108.59: warning: Gdk.Screen.get_monitor_at_point has been deprecated since 3.22

Both audience and presenter's screen are visible.

**Comment by oskar456** [comment]
Compiles without warnings and works both with vala 0.34 and 0.36. With vala 0.32 it does not compile (there's no `Gdk.Monitor` defined in `vapi/gdk-3.0.vapi`) so I guess minimum Vala version requirement should be raised from 0.26 to 0.34.

**Comment by fnevgeny** [comment]
@oskar456: it's strange about vala. I was under impression that as long as we require gtk-3.22 (where `Gdk.Monitor` was introduced), everything should be satisfied. Do you say vala 0.32 works with gtk-3.22 yet doesn't have `Gdk.Monitor` defined?

**Comment by AndreasBilke** [comment]
What surprised me is how did you managed to have multiple Vala versions in a normal setup? I would guess a normal installation has a fitting gtk Bala combination

**Comment by oskar456** [comment]
I'm not an expert on vala, in fact, it's my first time using it. With vala-0.32 the `cmake ..` phase goes on well:

```
$ cmake ..
-- The C compiler identification is GNU 7.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Vala: /usr/bin/valac-0.32 (found suitable version "0.32.1", minimum required is "0.26") 
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Checking for module 'gobject-2.0'
--   Found gobject-2.0, version 2.52.3
-- Checking for module 'gio-2.0'
--   Found gio-2.0, version 2.52.3
-- Checking for module 'gee-0.8'
--   Found gee-0.8, version 0.20.1
-- Checking for module 'poppler-glib>=0.22'
--   Found poppler-glib, version 0.62.0
-- Checking for module 'gtk+-3.0>=3.22'
--   Found gtk+-3.0, version 3.22.30
…
-- Configuring done
-- Generating done
-- Build files have been written to: /home/oskar/Downloads/pdfpc/build
```

In the `make` phase, it dies very early:
```
$ LANG=C make
[  3%] Generating classes/action/action_mapping.c, classes/action/link_action.c, classes/action/movie.c, classes/cache_status.c, classes/config_file_reader.c, classes/dbus_server.c, classes/drawings/drawing.c, classes/drawings/storage.c, classes/metadata/pdf.c, classes/metadata/slides_notes.c, classes/options.c, classes/presentation_controller.c, classes/renderer/cache/base.c, classes/renderer/cache/png/engine.c, classes/renderer/cache/png/item.c, classes/renderer/cache/png/persistent/engine.c, classes/renderer/cache/simple/engine.c, classes/renderer/pdf.c, classes/scaler.c, classes/timer_label.c, classes/view/behaviour/base.c, classes/view/behaviour/pdf_link.c, classes/view/pdf.c, classes/view/video.c, classes/window/fullscreen.c, classes/window/overview.c, classes/window/presentation.c, classes/window/presenter.c, interfaces/controllable.c, pdfpc.c
/home/oskar/Downloads/pdfpc/src/classes/window/fullscreen.vala:95.13-95.23: error: The type name `Gdk.Monitor' could not be found
            Gdk.Monitor monitor;
            ^^^^^^^^^^^
Compilation failed: 1 error(s), 0 warning(s)
make[2]: *** [src/CMakeFiles/pdfpc.dir/build.make:93: src/classes/action/action_mapping.c] Error 1
make[1]: *** [CMakeFiles/Makefile2:85: src/CMakeFiles/pdfpc.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
I use Gentoo Linux, I downloaded source tarballs of `vala-0.32.1` and `vala-0.34.16`, unpacked and compared:

```
$ grep -r 'Gdk\.Monitor'  /var/tmp/portage/dev-lang/vala-0.32.1/work/vala-0.32.1/
<nothing>
$ grep -r 'Gdk\.Monitor'  /var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/
/var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/vapi/gdk-3.0.vapi:             public unowned Gdk.Monitor? get_monitor (int monitor_num);
/var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/vapi/gdk-3.0.vapi:             public unowned Gdk.Monitor get_monitor_at_point (int x, int y);
/var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/vapi/gdk-3.0.vapi:             public unowned Gdk.Monitor get_monitor_at_window (Gdk.Window window);
/var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/vapi/gdk-3.0.vapi:             public unowned Gdk.Monitor? get_primary_monitor ();
/var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/vapi/gdk-3.0.vapi:             public signal void monitor_added (Gdk.Monitor monitor);
/var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/vapi/gdk-3.0.vapi:             public signal void monitor_removed (Gdk.Monitor monitor);
/var/tmp/portage/dev-lang/vala-0.34.16/work/vala-0.34.16/vapi/gtk+-3.0.vapi:            public void place_on_monitor (Gdk.Monitor monitor);

```

I traced down that such definitons were added by [this commit](https://github.com/GNOME/vala/commit/a3e2f588b428920c058f671656e50a856ef3b9ee) from 28 Apr 2016, which probably didn't make it to 0.32.1 released on 20 Jun 2016.

**Comment by oskar456** [comment]
> What surprised me is how did you managed to have multiple Vala versions in a normal setup? I would guess a normal installation has a fitting gtk Bala combination

Gentoo has [slotted ebuild](https://devmanual.gentoo.org/general-concepts/slotting/index.html) for vala. Each version's binary is suffixed by version number – `valac-0.32`, `valac-0.34`, `valac-0.36`. Also, there's no `valac` binary, so one has to patch `cmake/vala/FindVala.cmake`. This is normaly [done by the ebuild](https://github.com/gentoo/gentoo/blob/a0ee7a3e662615f0800b76fcbe1557871ab908ff/app-misc/pdfpc/pdfpc-4.2.1.ebuild#L34).

**Comment by fnevgeny** [comment]
@oskar456, thanks. Bumped Vala requirement to 0.34.

**Comment by fnevgeny** [comment]
> Also, there's no `valac` binary so one has to patch `cmake/vala/FindVala.cmake`

IMHO, rather a respective `eselect` module should be there...

**Comment by AndreasBilke** [comment]
Closed with #392
